core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -969,9 +1007,48 @@ wasi_poll_oneoff(wasm_exec_env_t exec_env, const wasi_subscription_t *in,
         || !validate_native_addr(nevents_app, sizeof(uint32)))
         return (wasi_errno_t)-1;
 
-    err = wasmtime_ssp_poll_oneoff(curfds, in, out, nsubscriptions, &nevents);
-    if (err)
-        return err;
+    __wasi_timestamp_t elapsed = 0;
+    const __wasi_timestamp_t timeout = get_timeout_for_poll_oneoff(
+                                 in, nsubscriptions),
+                             time_quant = 1e9;
+    const uint64 size_to_copy = nsubscriptions * sizeof(wasi_subscription_t);
+    __wasi_subscription_t *in_copy = NULL;
+    if (size_to_copy >= UINT32_MAX
+        || (size_to_copy
+            && !(in_copy = (__wasi_subscription_t *)wasm_runtime_malloc(
+                     (uint32)size_to_copy)))) {
+        return (wasi_errno_t)-1;
+    }
+    bh_memcpy_s(in_copy, size_to_copy, in, size_to_copy);
```

Comment:
Code is not easy to understand, had better change to:
```C
if (size_to_copy > 0) {
    if (size_to_copy >= UINT32_MAX
        || !(in_copy = (__wasi_subscription_t *)wasm_runtime_malloc(
                     (uint32)size_to_copy))) {
        return (wasi_errno_t)-1;
    }
    bh_memcpy_s(in_copy, size_to_copy, in, size_to_copy);
}
```

Reply:
or could we add at the beginning:
```c
if (nsubscriptions == 0) {
    return 0;
}
```

Reply:
That's good, in that case, here in_copy must be not NULL, and we don't need to check whether it is NULL when freeing it below.